cmd/testrunner/sensei.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/bottlepay/lightning-benchmark/sensei"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/routing/route"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+type senseiConnection struct {
+	conn   *grpc.ClientConn
+	pubKey route.Vertex
+	alias  string
+	client sensei.NodeClient
+}
+
+func getSenseiConnection(alias string) (*senseiConnection, error) {
+	logger := log.With("node", alias)
+
+	rpcHost := fmt.Sprintf("node-%v:5401", alias)
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+
+	logger.Infow("Attempting to connect to sensei")
+
+	var adminConn *grpc.ClientConn
+	err := tryFunc(
+		func() error {
+			var err error
+			adminConn, err = grpc.Dial(rpcHost, opts...)
+			if err != nil {
+				return err
+			}
+			return err
+		}, 60)
+	if err != nil {
+		return nil, err
+	}
+	defer adminConn.Close()
+
+	logger.Infow("Creating node")
+
+	adminClient := sensei.NewAdminClient(adminConn)
+
+	createResp, err := adminClient.CreateAdmin(context.Background(), &sensei.CreateAdminRequest{
+		Username:   "admin",
+		Alias:      alias,
+		Passphrase: "secret",
+		Start:      true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	pubkey, err := route.NewVertexFromStr(createResp.Pubkey)
+	if err != nil {
+		return nil, err
+	}
+
+	time.Sleep(time.Second)
+
+	mac := createResp.Macaroon
+
+	cred, err := NewMacaroonCredential(mac)
+	if err != nil {
+		return nil, err
+	}
+
+	opts = append(opts,
+		grpc.WithPerRPCCredentials(cred),
+	)
+
+	logger.Infow("Connecting with macaroon")
+
+	var conn *grpc.ClientConn
+	err = tryFunc(
+		func() error {
+			var err error
+			conn, err = grpc.Dial(rpcHost, opts...)
+			if err != nil {
+				return err
+			}
+			return err
+		}, 60)
+	if err != nil {
+		return nil, err
+	}
+
+	return &senseiConnection{
+		alias:  alias,
+		conn:   conn,
+		pubKey: pubkey,
+		client: sensei.NewNodeClient(conn),
+	}, nil
+}
+
+func (l *senseiConnection) PubKey() string {
+	return l.pubKey.String()
+}
+
+func (l *senseiConnection) Close() {
+	l.conn.Close()
+}
+
+func (l *senseiConnection) Connect(key, host string) error {
+	_, err := l.client.ConnectPeer(context.Background(), &sensei.ConnectPeerRequest{
+		NodeConnectionString: fmt.Sprintf("%v@%v", key, host),
+	})
+
+	return err
+}
+
+func (l *senseiConnection) NewAddress() (string, error) {
+	resp, err := l.client.GetUnusedAddress(
+		context.Background(),
+		&sensei.GetUnusedAddressRequest{},
+	)
+	if err != nil {
+		return "", err
+	}
+	return resp.Address, nil
+}
+
+func (l *senseiConnection) OpenChannel(peerKey string, amtSat int64,
+	pushAmtSat int64, private bool) error {
+
+	_, err := l.client.OpenChannels(context.Background(), &sensei.OpenChannelsRequest{
+		Channels: []*sensei.OpenChannelInfo{
+			{
+				AmtSatoshis: uint64(amtSat),
+				Public:      !private,
+
+				// TODO:
+				// NodePubkeyString:   peerKey,
+				// SpendUnconfirmed:   true,
```

Comment:
For other implementations, I open multiple channels from a single utxo by using the (unconfirmed) change. Maybe I should do a batch open with sensei?

Reply:
You can definitely do a batch channel open which will use bdk to construct a single on-chain tx for all the channels.  It relies on bdk wallet coin selection for the input utxos though.

Reply:
Ok. An if I open the channels one by one, will it be able to 'chain' the open requests using 0-conf change outputs?